Formalize .dart_tool/ (#4137)
dart-lang/site-www: src/tools/pub/package-layout.md

    style.css
{% endprettify %}

> \* The `.dart_tool/` directory exists after you've run `pub get`. 

Reviewer: I think it would be nice to keep this quick reference here as readers may not go further down the page.

Could add something like to reference that later section as well:

```markdown
To learn more, 
see [Project specific caching for tools](#project-specific-caching-for-tools).
```

Reviewer 2: Ahh, shoot, it's not letting me commit this suggestion.

> \* The `.dart_tool/` directory exists after you've run `pub get`. 

Reviewer 2: Running `pub get` creates the `.dart_tool` directory if none exists.
To learn more, 
see [Project specific caching for tools](#project-specific-caching-for-tools).

Reviewer: I've manually pushed a fix adding this back.

Unlike the scripts in `bin`, these are *not* for external users of the package.
If you have any of these, place them in a directory called `tool`.

## Project specific caching for tools

{{site.alert.info}}
  Do not check the `.dart_tool/` directory into source control.
  Instead keep `.dart_tool/` in `.gitignore`.
{{site.alert.end}}

> The `.dart_tool/` is created when you've run `dart pub get`.

Reviewer 3: "The `.dart_tool/` directory is created when you run `dart pub get`."

 (Add "directory", "when you've" -> "when you" - otherwise it should be "have been created when you've run")

> The directory used by various tooling for caching files specific to your project and/or local machine.

Reviewer 3: ... directory is used by various tools ...

"Various" suggests a number of things, but "tooling" is not countable.

Thus, the `.dart_tool/` directory should never be checked into source control, or copied between machines.

It is also generally safe to delete the `.dart_tool/` directory, though some tools might need recompute the cached information before they work again. 

> **Example:** The [dart pub get](https://dart.dev/tools/pub/cmd/pub-get) tool will dowload and extract dependencies to a global `$PUB_CACHE` directory, and then write a `.dart_tool/package_config.json` file mapping _package names_ to directories in the global `$PUB_CACHE` directory. The `.dart_tool/package_config.json` file is used by other tools, such as analyzer, compiler, etc. when they need to resolve statments such as `import "package:foo/foo.dart"`.

Reviewer 3: (Break long lines like this. It makes it easier to edit in the future.)

Add comma after "etc.".

When developing a tool that needs project specific caching, it can be advantagous to use the `.dart_tool/` directory because most users already have it ignored in `.gitignore`.
When caching files for your tool in a user's `.dart_tool/` directory, you should always use a unique subdirectory.

> To avoid collisions, subdirectories like `.dart_tool/<tool_package_name>/` are reserved for the package named `<tool_package_name>`.

Author: Package naming conventions are documented elsewhere (if they are documented at all). We could add it, but I doubt it's necessary, and we just risk failing to keep it updated.

Unlike the scripts in `bin`, these are *not* for external users of the package.
If you have any of these, place them in a directory called `tool`.

## Project specific caching for tools

{{site.alert.info}}
  Do not check the `.dart_tool/` directory into source control.
  Instead keep `.dart_tool/` in `.gitignore`.
{{site.alert.end}}

The `.dart_tool/` directory is created when you run `dart pub get`.
The directory is used by various tools for caching files specific to your project and/or local machine.
The `.dart_tool/` directory should never be checked into source control, or copied between machines.

> It is also generally safe to delete the `.dart_tool/` directory, though some tools might need recompute the cached information before they work again. 

Reviewer 4: Should we strengthen this, so people using the `.dart_tool` directory know what the contract is: "Your files might get deleted at any time".

It is safe to delete the `.dart_tool/` directory, though some tools might need recompute the cached information before they work again.

Reviewer: ```suggestion
It is safe to delete the `.dart_tool/` directory,
but some tools might need recompute the cached files
before they work again. 
```

Reviewer 2: I believe this is addressed.

**Example:** The [dart pub get](https://dart.dev/tools/pub/cmd/pub-get) tool
will dowload and extract dependencies to a global `$PUB_CACHE` directory,
and then write a `.dart_tool/package_config.json` file mapping _package names_
to directories in the global `$PUB_CACHE` directory.
The `.dart_tool/package_config.json` file is used by other tools,
such as analyzer, compiler, etc, when they need to resolve statments

> such as `import "package:foo/foo.dart"`.

Reviewer: These changes look good. I've also adjusted the format and link to `dart pub get` in the suggestion.

> such as `import "package:foo/foo.dart"`.

Reviewer: These changes look great, thanks. I've also adjusted the format and link to `dart pub get` in the suggestion.

When developing a tool that needs project specific caching, it can be
advantagous to use the `.dart_tool/` directory because most users already
ignore the directory using `.gitignore`.
When caching files for your tool in a user's `.dart_tool/` directory,
you should always use a unique subdirectory.
To avoid collisions, subdirectories of the form `.dart_tool/<tool_package_name>/`
are reserved for the package named `<tool_package_name>`.

> If your tool is not distributed through [pub.dev](https://pub.dev), you may publish a placeholder package in order to reserve the reserve a unique name.

Reviewer: Thanks for cleaning this up. I've also adjusted the pub link.

> If your tool is not distributed through [pub.dev](https://pub.dev), you may publish a placeholder package in order to reserve the reserve a unique name.

Reviewer: Thanks for cleaning this up. I've adjusted some line breaks and the pub link in the provided suggestion.
